Fix state hash mismatch error that caused by ignored MPT error

peterlimg opened the pull request:
This PR was created to fix the new detected 'state hash mismatch' error that caused the chain stuck. But after that, new issues were detected during the loadtests, hence all fixes were also added here. Anyway, the core fixes and changes are as shown below.

## Fixes
- Fixed the state hash mismatch error that caused the chain stuck. 'node not found' error was not returned to upper layer that caused the txn executor see the error as charageable error, while other miners that has full state would not see this error, therefore, they will see different block state, and cause chain stuck.
- Fixed the slow finalizing process on sharders, it took about 1.2s to store the miner rewards update events to db. Optimized to do bulk update for all the events in one sql execution. Also, maximize the concurrency on finalizing a blocks. 
- Fixed the partition save bug
- Fixed a sharder blocks sync bug that caused the sharder stuck
- Fixed allocation challenges stats bug

## Changes
- Changed the stake pools to use `txn.ClientID` as key. Previously, we use txn hash as key, which does not make sense at all, as each provider has limited size of delegate/stake pools, if we use txn hash as key, them we can only have max limit size of txns to stake the pools. For example, if miner one has max of 200 delegate pools, if we use txn hash as pool key, then we will only allow 200 stake transactions, then no more staking could be allowed. While, after the fix, we use client id as pool id, which means, as long as the client can stake the miner once, it then can stake as many times as it wants, and new staked tokens will all be added up to the same pool. This would also simplify the clients usage, basically all the pool_id params will be removed from the `zboxcli`, `zwalletcli`.
- Minor refactoring for fixing potential goroutine leaks.

## Need to be mentioned in CHANGELOG.md?

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- gosdk: https://github.com/0chain/gosdk/pull/556
- system_test: https://github.com/0chain/system_test/pull/380
- zboxcli: https://github.com/0chain/zboxcli/pull/291
- zwalletcli: https://github.com/0chain/zwalletcli/pull/172
- loadTest: https://github.com/0chain/loadTest/pull/53

peterlimg committed:
Should not ignore error for getBlobbers

peterlimg committed:
Fix missing MPT error

peterlimg committed:
Fix potential goroutine leak

peterlimg committed:
Fix hidden MPT error (though not being used)

peterlimg committed:
Minor refactor

peterlimg committed:
Fix missed MPT error

peterlimg committed:
Return when eventdb is nil, we should always return if error happens

peterlimg committed:
Minor refactor

peterlimg committed:
Fix state hash mismatch

peterlimg committed:
Fix unit test

peterlimg committed:
Fix sharder rounds saving on finalizing

- Do not save blocks and rounds asynchronously
- Clone round to avoid finalized round being changed before saving to db

peterlimg committed:
Speed up finalizing process on sharders

peterlimg committed:
Panic if store finaliz block/round to db failed

peterlimg committed:
Process finalized block concurrently

peterlimg committed:
Debug finalize block duration

peterlimg committed:
Debug event db metrics

peterlimg committed:
Restrict to one stake pool per use for each provider

peterlimg committed:
Fix block clone missing changes

peterlimg committed:
One stakepool per miner

peterlimg committed:
Use txn.ClientID as stake pool id

peterlimg committed:
Allow user to collect rewards for staked delegate pool

peterlimg committed:
Do not assign currency operation result directly

peterlimg committed:
Debug block fees and rewards

peterlimg committed:
Fix panic pool status panic

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Bulk update rewards

peterlimg committed:
Debug reward update

peterlimg committed:
Fix bulk reward delegate pool update

peterlimg committed:
Fix empty delegate pool update

Kishan-Dhakan committed:
Merge branch 'staging' into fix/state-hash-mismatch-3

peterlimg committed:
Fix partitions save

peterlimg committed:
Fix useless user stake pool provider type

peterlimg committed:
Remove unused function

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Fix 'previous block state not computed' error

Block in chain.blocks and round are different, so it is possible that the block in the round is computed, but not synced to chain.blocks. Since next round will get previous block from chain.blocks, then it will see the block is not computed.

Kishan-Dhakan committed:
Merge branch 'staging' into fix/state-hash-mismatch-3

peterlimg committed:
Fix delegateFromDelegate sc to use clientid as pool id

peterlimg committed:
Fix unit tests

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Clean up debug logs

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg committed:
Fix sharder block sync issue

peterlimg committed:
Clean up logs

peterlimg committed:
Fix lint error

peterlimg committed:
Fix incorrect updating of TotalChallenges

peterlimg committed:
Reduce events buffer size and add more event metrics

peterlimg committed:
Fix unit tests

peterlimg committed:
Fix unit tests

peterlimg committed:
Fix bugs in alloc stats when finalize and cancel

peterlimg committed:
Fix unit tests in challenge stats

peterlimg committed:
Debug empty total challenges

peterlimg committed:
Fix incorrect error report of 'empty total challenges'

dabasov commented on code/go/0chain.net/chaincore/chain/protocol_block.go:
```diff
@@ -403,26 +418,62 @@ func (c *Chain) finalizeBlock(ctx context.Context, fb *block.Block, bsh BlockSta
+		zap.Any("duration", time.Since(ts)))
+}
+
+type waitGroupSync struct {
```
we might want to mow it to common.context? it is reasonable to have concurrency structures all in one place

peterlimg replied:
Perhaps, if needed, we can move to a common package.

dabasov commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -1279,21 +1315,32 @@ func (sc *StorageSmartContract) finalizedPassRates(alloc *StorageAllocation) ([]
-		ba.Stats.TotalChallenges = ba.Stats.FailedChallenges + ba.Stats.SuccessChallenges
-		if ba.Stats.TotalChallenges == 0 {
+
+		baTotal := ba.Stats.FailedChallenges + ba.Stats.SuccessChallenges
```
should we update ``ba.Stats.TotalChallenges`` with this value?

peterlimg replied:
No, the `ba.Stats.TotalChallenges` will only be updated when new challenges are generated, update here could lead to wrong stats.

dabasov reviewed (approved):
lgtm

peterlimg committed:
Merge branch 'staging' into fix/state-hash-mismatch-3

peterlimg committed:
Install mockery

peterlimg committed:
Fix unit test errors

peterlimg committed:
Debug stake pool lock fail

peterlimg committed:
Merge branch 'staging' into fix/state-hash-mismatch-3

peterlimg committed:
Fix getStakePoolsByPools/getBlobbersByIDs

return error if any immediately instead of continue and return the ones that could be acquired

peterlimg committed:
Fix swagger

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | staging  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | one-pool  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | one-pool  |
| gosdk  | one-pool  |
| zboxcli  | one-pool  |
| zwalletcli  | one-pool  |
| 0chain  | fix/state-hash-mismatch-3  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |
